feat: Supporting managed types in NetworkVariable [MTT-4594]

ShadauxCat opened the pull request:
Also changes NetworkVariable to deserialize INetworkSerializable types in-place - this is done to avoid having a GC alloc every time the value changes. There will only be a GC alloc if it receives an updated value and the current value is `null`.

Managed types are supported in two ways:

1. By making them `INetworkSerializable`, or
2. By assigning the delegates `UserNetworkVariableSerialization<T>.WriteValue` and `UserNetworkVariableSerialization<T>.ReadValue`

If a managed type is passed in without meeting one of those criteria (or any unsupported type), then an exception will be thrown the first time the NetworkVariable is read or written. That timing is to ensure no weird behavior with exceptions being thrown before user code is able to assign the delegates.

## Changelog

- Changed: `NetworkVariable<>` now supports managed `INetworkSerializable` types, as well as other managed types with serialization/deserialization delegates registered to `UserNetworkVariableSerialization<T>.WriteValue` and `UserNetworkVariableSerialization<T>.ReadValue`
- Changed: `NetworkVariable<>` and `BufferSerializer<BufferSerializerReader>` now deserialize `INetworkSerializable` types in-place, rather than constructing new ones.

## Testing and Documentation

- Includes integration tests.
- Includes edits to existing public API documentation.

## Generated ILPP output:

The ILPP generates two function calls for each type it finds being used in a NetworkVariable - one to set a serializer for it, one to set an equality operator for it. The methods being called are all found in `NetworkVariableSerialization.cs` (and all added in this review so they're easy to see there). The reason ILPP is used is because the methods being called for various types have constraints (for example, `FastBufferWriter.WriteUnmanagedSafe" has an `unmanaged` constraint). Since that constraint was removed from `NetworkVariable`, it's no longer possible to call that method, nor even to initialize the serializer as there's no way to cast a type to say "this type is definitely unmanaged, compiler" in order to create an instance of that class, and using reflection to create it doesn't work in IL2CPP... but initializing a delegate based on that type does work IF the type is known at compile time in the location it's being initialized. ILPP is used to call those initialization functions to save our users the headache of having to remember to initialize all their network variable types - without ILPP, they'd have to call all the initialization manually.

In the future, Code Generators will offer better solutions for this.

![image](https://user-images.githubusercontent.com/284434/193914428-abdb72c0-d612-45f8-8d4c-3b3babb302a2.png)

ShadauxCat committed:
feat: Supporting managed types in NetworkVariable.

ShadauxCat committed:
changelog

ShadauxCat committed:
Removed debug log line

ShadauxCat committed:
Fixed compile error... apparently HashCode.Combine doesn't exist in all versionf of unity, and it's not used anyway so I just set it to return 0.

0xFA11 commented on com.unity.netcode.gameobjects/CHANGELOG.md:
```diff
@@ -16,6 +16,8 @@ Additional documentation and release notes are available at [Multiplayer Documen
 
 - The debug simulator in `UnityTransport` is now non-deterministic. Its random number generator used to be seeded with a constant value, leading to the same pattern of packet drops, delays, and jitter in every run. (#2196)
+- `NetworkVariable<>` now supports managed INetworkSerializable types, as well as other managed types with serialization/deserialization delegates registered to `UserNetworkVariableSerialization<T>.WriteValue` and `UserNetworkVariableSerialization<T>.ReadValue` (#2219)
+- `NetworkVariable<>` and `BufferSerializer<BufferSerializerReader>` now deserialize `INetworkSerializable` types in-place, rather than constructing new ones. This makes using native collections (such as `NativeArray<>`) easier as they can be reused once they're constructed rather than having to construct new ones and free old ones on every update, and prevents unnecessary GC allocations with man ...
```
<3

NoelStephensUnity commented on com.unity.netcode.gameobjects/CHANGELOG.md:
```diff
@@ -16,6 +16,8 @@ Additional documentation and release notes are available at [Multiplayer Documen
 ### Changed
 
 - The debug simulator in `UnityTransport` is now non-deterministic. Its random number generator used to be seeded with a constant value, leading to the same pattern of packet drops, delays, and jitter in every run. (#2196)
+- `NetworkVariable<>` now supports managed INetworkSerializable types, as well as other managed types with serialization/deserialization delegates registered to `UserNetworkVariableSerialization<T>.WriteValue` and `UserNetworkVariableSerialization<T>.ReadValue` (#2219)
```
total minor "nit-pick":
`INetworkSerializable`

NoelStephensUnity reviewed (approved):
Minor nit with changelog.
Everything else LGTM!

ShadauxCat commented:
Unfortunate... this code fails on consoles. I'm going to close this for now and will re-open once I get consoles working again.

ShadauxCat committed:
Split NetworkVariable into NetworkVariable and ManagedNetworkVariable because C# generics just aren't flexible to do it in one class without reflection, and I couldn't find a way to make the reflection work with il2cpp.

ShadauxCat commented:
Split NetworkVariable into NetworkVariable and ManagedNetworkVariable because C# generics just aren't flexible to do it in one class without reflection, and I couldn't find a way to make the reflection work with il2cpp.

ShadauxCat committed:
Updated changelog

ShadauxCat committed:
Changelog consistency

ShadauxCat commented:
Jesse has asked me to spend some more time trying to find a way around needing a separate class. I'm closing this again because it shouldn't be reviewed until it's finalized.

ShadauxCat committed:
Returned to a single NetworkVariable class. Removed all reflection and went back to an ILPP-based approach.

Limitation to this approach is that only NetworkVariables which are fields of NetworkBehaviours will get initialized... but as we don't really support any other usage of NetworkVariables, this seems like an acceptable limitation.

Also upgraded the project to 2020.3.40, as older versions fail to compile some of the new tests due to IL2CPP bugs.

ShadauxCat commented:
Back to only one type.

ShadauxCat committed:
Merge branch 'develop' into feat/managed_types_in_networkvariable

ShadauxCat committed:
Standards

ShadauxCat committed:
Merge remote-tracking branch 'origin/feat/managed_types_in_networkvariable' into feat/managed_types_in_networkvariable

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/Serialization/FastBufferReader.cs:
```diff
@@ -1310,5 +1323,24 @@ public unsafe void ReadValueSafe<T>(out T value, FastBufferWriter.ForFixedString
+
+        /// <summary>
+        /// Read a FixedString value.
+        ///
```
nit-pick?
good place for a `<remarks></remarks>` as opposed to the black comment line?

ShadauxCat replied:
For a...?

NoelStephensUnity replied:
Sorry, forgot to mark the remarks as code...

NoelStephensUnity commented on testproject/Packages/manifest.json:
```diff
@@ -1,15 +1,15 @@
-    "com.unity.ide.visualstudio": "2.0.11",
-    "com.unity.ide.vscode": "1.2.3",
-    "com.unity.mathematics": "1.2.1",
+    "com.unity.addressables": "1.18.19",
```
Are the manifest and packages json file modifications part of this PR?

ShadauxCat replied:
Yes, they are intended to be here. :)

NoelStephensUnity commented on testproject/ProjectSettings/ProjectVersion.txt:
```diff
@@ -1,2 +1,2 @@
-m_EditorVersion: 2020.3.12f1
-m_EditorVersionWithRevision: 2020.3.12f1 (b3b2c6512326)
+m_EditorVersion: 2020.3.40f1
```
Recalling the conversations in the thread about this PR, there was a mention of this version so I am guessing the manifest and packages-lock updates are based on this?

ShadauxCat replied:
Exactly. The new tests fail to compile on earlier versions.

NoelStephensUnity reviewed (approved):
Would definitely get @ThusSpokeNomad to look over this as well.
Minor nit on possibly using remarks for a public API method.
Just commented for verification on the ProjectVersion and associated json files that were changed.
Other than that it LGTM!

0xFA11 reviewed (changes requested):
Haven't had the full code-review yet but I noticed the description is out-of-date:

> Changed: Added a new ManagedNetworkVariable<> with support for managed INetworkSerializable types

Do you mind updating the description to reflect up-to-date changes? (Also perhaps CHANGELOG.md too?)

ShadauxCat committed:
Renamed a function and forgot to rename it in a comment that references it

0xFA11 commented on com.unity.netcode.gameobjects/Editor/CodeGen/NetworkBehaviourILPP.cs:
```diff
@@ -172,6 +319,9 @@ public override ILPostProcessResult Process(ICompiledAssembly compiledAssembly)
         private const string k_ServerRpcParams_Receive = nameof(ServerRpcParams.Receive);
         private const string k_ServerRpcReceiveParams_SenderClientId = nameof(ServerRpcReceiveParams.SenderClientId);
 
+        // CodeGen cannot reference the collections assembly to do a typeof() on it due to a bug that causes that to crash.
+        private const string k_INativeListBool_FullName = "Unity.Collections.INativeList`1<System.Byte>";
```
:(

0xFA11 commented on com.unity.netcode.gameobjects/Tests/Runtime/Components/NetworkVariableTestComponent.cs:
```diff
@@ -1,8 +1,161 @@
+
+        public override int GetHashCode()
+        {
+            return 0;
```
nit
```suggestion
            return Str.GetHashCode() ^ Int.GetHashCode() ^ InMemoryValue.GetHashCode();
```

0xFA11 reviewed (approved):
Left a few nits but overall it's looking good.

0xFA11 committed:
Merge branch 'develop' into feat/managed_types_in_networkvariable

0xFA11 committed:
Update com.unity.netcode.gameobjects/Tests/Runtime/Components/NetworkVariableTestComponent.cs

0xFA11 committed:
chore: standards fix, add meta files

0xFA11 committed:
Merge branch 'develop' into feat/managed_types_in_networkvariable

ShadauxCat committed:
Merge branch 'develop' into feat/managed_types_in_networkvariable
